pipeline-maven/src/main/java/org/jenkinsci/plugins/pipeline/maven/publishers/CoveragePublisher.java

```diff
@@ -0,0 +1,268 @@
+                            Level.FINE,
+                            "Ignore execution event with missing 'build' child:" + XmlUtils.toString(event));
+                continue;
+            }
+
+            Element pluginElt = XmlUtils.getUniqueChildElement(event, "plugin");
+            Element outputDirectoryElt = XmlUtils.getUniqueChildElementOrNull(pluginElt, "outputDirectory");
+            Element projectElt = XmlUtils.getUniqueChildElement(event, "project");
+            MavenArtifact mavenArtifact = XmlUtils.newMavenArtifact(projectElt);
+            MavenSpyLogProcessor.PluginInvocation pluginInvocation = XmlUtils.newPluginInvocation(pluginElt);
+
+            if (outputDirectoryElt == null) {
+                listener.getLogger()
+                        .println("[withMaven] No <outputDirectory> element found for <plugin> in "
+                                + XmlUtils.toString(event));
+                continue;
+            }
+            String outputDirectory = XmlUtils.resolveMavenPlaceholders(outputDirectoryElt, projectElt);
+            if (outputDirectory == null) {
+                listener.getLogger()
+                        .println(
+                                "[withMaven] could not resolve placeholder '${project.build.directory}' or '${project.reporting.outputDirectory}' or '${basedir}' in "
+                                        + XmlUtils.toString(event));
+                continue;
+            }
+
+            String resultFile = outputDirectory + File.separator + reportFilename;
+            patterns.add(XmlUtils.getPathInWorkspace(resultFile, workspace));
+
+            listener.getLogger()
+                    .println("[withMaven] coveragePublisher - Archive " + name + " analysis results for Maven artifact "
+                            + mavenArtifact.toString() + " generated by " + pluginInvocation + ", resultFile: "
+                            + resultFile);
+        }
+
+        if (patterns.isEmpty() && StringUtils.isBlank(extraPattern)) {
+            return null;
+        }
+
+        StringBuilder patternsAsString = new StringBuilder();
+        patternsAsString.append(patterns.stream().collect(Collectors.joining(",")));
+        if (StringUtils.isNotBlank(extraPattern)) {
+            if (!patterns.isEmpty()) {
+                patternsAsString.append(",");
+            }
+            patternsAsString.append(extraPattern);
+        }
+
+        CoverageTool tool = new CoverageTool();
+        tool.setParser(parser);
+        tool.setPattern(patternsAsString.toString());
+        return tool;
+    }
+
+    private SourceCodeDirectory toSourceDirectory(Element e) {
+        Element buildElement = XmlUtils.getUniqueChildElementOrNull(e, "project", "build");
+        if (buildElement == null) {
+            return null;
+        }
+        return new SourceCodeDirectory(buildElement.getAttribute("sourceDirectory"));
+    }
+
+    @Symbol("coveragePublisher")
+    @Extension
```

Comment:
Perhaps this was intended to be an optional extension?

Reply:
https://www.jenkins.io/doc/developer/plugin-development/optional-dependencies/ explains it better than I can

Reply:
Fixed in https://github.com/jenkinsci/pipeline-maven-plugin/pull/931 and https://github.com/jenkinsci/pipeline-maven-plugin/pull/933